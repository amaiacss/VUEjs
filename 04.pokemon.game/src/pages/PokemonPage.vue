<template>
    <h1 v-if="!pokemon" >Espere por favor.....</h1>
    <div v-else>
        <h1>¿Quién es este pokémon</h1>
        <PokemonPicture 
            :pokemonId="pokemon.id" 
            :showPokemon="showPokemon"
        />
        <PokemonOptions 
                :pokemons="pokemonArr"
                @selection="checkAnswer($event)" 
        /> <!--Estar escuchando el evento click de PokemonOptions llamado selection, el event es el valor que está emitiendo el selection desde PokemonOptions. Si como es en este caso, sólo se recibe un argumento/parámetro, no es necesario escribirlo, estaría implícito, automaticamente la funcion lo recibe-->    
        <template v-if="showAnswer" >
            <h2 class="fade-in">{{ message }}</h2>
            <button @click="newGame">Nuevo Juego</button>   
        </template>
    </div>
    
</template>

<script>
    import PokemonPicture from '../components/PokemonPicture.vue'
    import PokemonOptions from '@/components/PokemonOptions.vue'

    import getPokemonOptions from '@/helpers/getPokemonOptions'

    console.log( getPokemonOptions() )

    export default {
        components: { PokemonPicture, PokemonOptions },
        data() {
            return {
               pokemonArr: [],
               pokemon: null,
               showPokemon: false,
               showAnswer: false,
               message: '',
            }
        },
        methods: {
            async mixPokemonArray() {
                this.pokemonArr = await getPokemonOptions() // Esta funcion es una promesa hay que esperar que se resuelva
                // Generar un numero del 0 al 3 al azar, que son las posiciones del array de Pokemons
                const rndInt = Math.floor( Math.random() * 4 )
                this.pokemon = this.pokemonArr[ rndInt ]
            },
            checkAnswer( selectedId ) {
                this.showPokemon = true
                this.showAnswer = true
                if( selectedId === this.pokemon.id ) {
                    this.message = `Correcto, ${ this.pokemon.name }`
                } else {
                    this.message = `Ooops, era ${ this.pokemon.name }`
                }
            },
            newGame() {
                // reinicializar
                this.showPokemon = false
                this.showAnswer =false
                this.pokemonArr = []
                this.pokemon = null
                this.mixPokemonArray()
            }
        },
    // LIFECYCLE HOOKS, ciclo de vida del componente, con mount cuando el componente esté montado , ejecuta lo que está dentro de este
        mounted() {
            this.mixPokemonArray() // llamada al método, lo que se recibe aquí es realmente un objeto, una referencia reactiva del array, en este caso. Tiene los getters y setters......
        }
    }
</script>

