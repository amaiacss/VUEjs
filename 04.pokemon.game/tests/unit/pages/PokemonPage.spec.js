import { shallowMount, mount} from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
// importar array para la prueba 3
import { mockPokemons } from '../mocks/pokemons.mock'

describe('Pokemon Page Component', () => {

    let wrapper
    beforeEach(() => {
        wrapper = shallowMount( PokemonPage )
    })

    test('Debe hacer math con el snapshot', () => {
        expect( wrapper.html()).toMatchSnapshot() 
    })

    // Evaluamos que cuando el componente se monta llamada a el método
    test('Debe llamar mixPokemonArray al montar', async () => {
        // espiamos los métodos del PokemonPage, el segundo argumento es el método que quiero que espie
        const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray' )
        // volvemos a montar el componente, porque en el before lo monta, y para cuando quiere construir el spy en este test, el mount ya se ha disparado y no lo llega a espiar, por lo que lo volvemos a montar el componente y así lo espía y lo puede testear
        shallowMount( PokemonPage )
        expect( mixPokemonArraySpy ).toHaveBeenCalled()
    })
    test('Debe hacer match con el snapshot cuando cargan los pokemons', () => {
        //montar el componente y mandarle la data con los datos como quiero hacer la prueba
        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: mockPokemons ,
                    pokemon:mockPokemons[0],
                    showPokemon:false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect( wrapper.html() ).toMatchSnapshot()

    })

    test('Debe mostrar los componentes de PokemonPicture y PokemonOptions', () => {
        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: mockPokemons ,
                    pokemon:mockPokemons[0],
                    showPokemon:false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect( wrapper.find('pokemon-picture-stub').exists() ).toBeTruthy()
        expect( wrapper.find('pokemon-options-stub').exists() ).toBeTruthy()

        expect( wrapper.find('pokemon-picture-stub').attributes('pokemonid') ).toBe('1')
        expect( wrapper.find('pokemon-options-stub').attributes('pokemons') ).toBeTruthy()
    })
})