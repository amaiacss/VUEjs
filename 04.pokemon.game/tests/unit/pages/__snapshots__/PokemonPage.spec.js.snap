// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon Page Component Debe hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>¿Quién es este pokémon</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--Estar escuchando el evento click de PokemonOptions llamado selection, el event es el valor que está emitiendo el selection desde PokemonOptions. Si como es en este caso, sólo se recibe un argumento/parámetro, no es necesario escribirlo, estaría implícito, automaticamente la funcion lo recibe-->
  <!--v-if-->
</div>
`;

exports[`Pokemon Page Component Debe hacer math con el snapshot 1`] = `<h1>Espere por favor.....</h1>`;
